### PR TITLE
♻️ Refactor docking to use rects for targets

### PR DESCRIPTION
**Pull request by alanorozco**
- Simplifies types/style.
- Removes vertical axis edge logic (it was incomplete and not exposed to user.)
- Extracts utilities for conciseness (particularly in `math.js`).
- Consolidates element slot vs. dynamic corner behavior with a standard rectangle concept.

### CODE REVIEW

File: extensions/amp-video-docking/0.1/def.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** @enum {string} */
+export const RelativeX = {LEFT: 'left', RIGHT: 'right'};
```
Comment (wassgha) [inline extensions/amp-video-docking/0.1/def.js]: Let's use int enums?

File: extensions/amp-video-docking/0.1/def.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
```
Comment (wassgha) [inline extensions/amp-video-docking/0.1/def.js]: 2020

File: extensions/amp-video-docking/0.1/amp-video-docking.js [hunk 3]
```diff
@@ -54,53 +70,90 @@ import {
...
+ * @struct @typedef {{
+ *   type: TargetType,
+ *   rect: !LayoutRectDef,
+ * }}
+ */
+let TargetDef;
+
+// Function should ideally be in `dom.js`, but moving it causes a bunch of ads
+// tests to fail, for some reason.
+// TODO(alanorozco): Move.
```
Comment (wassgha) [inline extensions/amp-video-docking/0.1/amp-video-docking.js]: Is this still the case?

File: extensions/amp-video-docking/0.1/viewport-rect.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
```
Comment (wassgha) [inline extensions/amp-video-docking/0.1/viewport-rect.js]: 2020

File: extensions/amp-video-docking/0.1/amp-video-docking.js [hunk 5]
```diff
@@ -936,16 +784,14 @@ export class VideoDocking {
   }
 
   /**
-   * @param {!../../../src/video-interface.VideoOrBaseElementDef} video
-   * @param {!DockTargetDef} target
+   * @param {!VideoOrBaseElementDef} video
+   * @param {!TargetDef} target
```
Comment (wassgha) [inline extensions/amp-video-docking/0.1/amp-video-docking.js]: I think `DockTargetDef` is less ambiguous here, any reason why we're dropping the `Dock`?

File: extensions/amp-video-docking/0.1/amp-video-docking.js [hunk 6]
```diff
@@ -487,7 +502,7 @@ export class VideoDocking {
     }
 
     const scrollDirection =
```
Comment (wassgha) [inline extensions/amp-video-docking/0.1/amp-video-docking.js]: Is this `const` needed?

File: extensions/amp-video-docking/0.1/def.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** @enum {string} */
+export const RelativeX = {LEFT: 'left', RIGHT: 'right'};
+
+/** @enum {string} */
+export const RelativeY = {TOP: 'top', BOTTOM: 'bottom'};
```
Comment (wassgha) [inline extensions/amp-video-docking/0.1/def.js]: Looks like this is only used for scroll direction anymore so can we rename it to `ScrollDirection`?

Comment (alanorozco) [reply extensions/amp-video-docking/0.1/def.js]: `DirectionX` and `DirectionY` as compromise ;)

File: extensions/amp-video-docking/0.1/amp-video-docking.js (hunk 6 above)
Comment (alanorozco) [reply extensions/amp-video-docking/0.1/amp-video-docking.js]: yeah, it's assigned as member read later.

File: extensions/amp-video-docking/0.1/amp-video-docking.js (hunk 3 above)
Comment (alanorozco) [reply extensions/amp-video-docking/0.1/amp-video-docking.js]: will address later

Comment (alanorozco) [reply extensions/amp-video-docking/0.1/amp-video-docking.js]: this util is actually no longer needed

File: extensions/amp-video-docking/0.1/amp-video-docking.js [hunk 8]
```diff
@@ -54,52 +68,80 @@ import {
...
 
-/** @enum */
-export const RelativeX = {LEFT: 0, RIGHT: 1};
+/** Width/height of placeholder icon when the inline box is small. */
+export const PLACEHOLDER_ICON_SMALL_WIDTH = 32;
 
-/** @enum */
-export const RelativeY = {TOP: 0, BOTTOM: 1};
+/** Margin applied to placeholder icon when the inline box is large. */
+export const PLACEHOLDER_ICON_SMALL_MARGIN = 20;
```
Comment (wassgha) [inline extensions/amp-video-docking/0.1/amp-video-docking.js]: Can these margins be combined with the breakpoints object below?

File: extensions/amp-video-docking/0.1/amp-video-docking.js [hunk 9]
```diff
@@ -54,52 +68,80 @@ import {
 
 const TAG = 'amp-video-docking';
 
-/** @visibleForTesting @const {number} */
-export const MARGIN_MAX = 30;
+/** Ratio to viewport width to use when docked to corner. */
+export const CORNER_WIDTH_RATIO = 0.3;
```
Comment (wassgha) [inline extensions/amp-video-docking/0.1/amp-video-docking.js]: Where is this coming from?

File: extensions/amp-video-docking/0.1/amp-video-docking.js [hunk 10]
```diff
@@ -54,52 +68,80 @@ import {
...
+/** @visibleForTesting @const {string} */
+export const BASE_CLASS_NAME = 'i-amphtml-video-docked';
 
 /** @enum {string} */
 export const Actions = {DOCK: 'dock', UNDOCK: 'undock'};
 
-/** @visibleForTesting */
-export const DOCKED_TO_CORNER_SIZING_RATIO = 0.3;
+/** @enum {string} */
+export const DockTargetType = {
```
Comment (wassgha) [inline extensions/amp-video-docking/0.1/amp-video-docking.js]: Do we expect any more dock target types? If not, would a Boolean be sufficient here? `isSlotTaget`?

File: extensions/amp-video-docking/0.1/viewport-rect.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+ * limitations under the License.
+ */
+import {LayoutRectDef, layoutRectLtwh} from '../../../src/layout-rect';
+
+/**
+ * @param {function():T} get
+ * @return {{get: function():T, configurable: boolean, enumerable: boolean}}
+ * @template T
+ */
+const readonlyGetterProp = get => ({
```
Comment (wassgha) [inline extensions/amp-video-docking/0.1/viewport-rect.js]: Hmm can you elaborate on why this is needed?

File: extensions/amp-video-docking/0.1/amp-video-docking.js (hunk 9 above)
Comment (alanorozco) [reply extensions/amp-video-docking/0.1/amp-video-docking.js]: It's static and not configurable, it was defined already under a different name.

File: extensions/amp-video-docking/0.1/amp-video-docking.js (hunk 8 above)
Comment (alanorozco) [reply extensions/amp-video-docking/0.1/amp-video-docking.js]: Nope, they're different concepts applied to different boxes (e.g. inline video area vs. dock area).

File: extensions/amp-video-docking/0.1/amp-video-docking.js (hunk 10 above)
Comment (alanorozco) [reply extensions/amp-video-docking/0.1/amp-video-docking.js]: Hmm, there's no plan to have more targets but I'd rather keep the disambiguation at the same abstract level via a type.

File: extensions/amp-video-docking/0.1/viewport-rect.js (hunk 11 above)
Comment (alanorozco) [reply extensions/amp-video-docking/0.1/viewport-rect.js]: Viewport box is not static, so in order to get a `Rect`-like box using the viewport, you need readonly getters :)

Comment (alanorozco) [reply extensions/amp-video-docking/0.1/viewport-rect.js]: If the question is why I added this general concept, the answer is internal simplicity. It's a single object and it's quite cheap.

**Pull request merged**